src/sb16.js

```diff
@@ -0,0 +1,1607 @@
+{
+    if(this.command & (1 << 3))
+    {
+        // Analogue to digital not implemented.
+        this.dsp_default_handler();
+        return;
+    }
+    var mode = this.write_buffer.shift();
+    this.dma_irq = SB_IRQ_8BIT;
+    this.dma_channel = this.dma_channel_8bit;
+    this.dma_autoinit = !!(this.command & (1 << 2));
+    this.dsp_signed = !!(mode & (1 << 4));
+    this.dsp_stereo = !!(mode & (1 << 5));
+    this.dsp_16bit = false;
+    this.dma_transfer_size_set();
+    this.dma_transfer_start();
+});
+
+// Pause 8-bit DMA mode digitized sound I/O.
+register_dsp_command([0xD0], 0);
+
+// Turn on speaker.
+// Documented to have no effect on SB16.
+register_dsp_command([0xD1], 0, function()
+{
+    this.dummy_speaker_enabled = true;
+});
+
+// Turn off speaker.
+// Documented to have no effect on SB16.
+register_dsp_command([0xD3], 0, function()
+{
+    this.dummy_speaker_enabled = false;
+});
+
+// Continue 8-bit DMA mode digitized sound I/O.
+register_dsp_command([0xD4], 0);
+
+// Pause 16-bit DMA mode digitized sound I/O.
+register_dsp_command([0xD5], 0);
+
+// Continue 16-bit DMA mode digitized sound I/O.
+register_dsp_command([0xD6], 0);
+
+// Get speaker status.
+register_dsp_command([0xD8], 0, function()
+{
+    this.read_buffer.clear();
+    this.read_buffer.push(this.dummy_speaker_enabled * 0xFF);
+});
+
+// Exit 16-bit auto-init DMA mode digitized sound I/O.
+// Exit 8-bit auto-init mode digitized sound I/O.
+register_dsp_command([0xD9, 0xDA], 0, function()
+{
+    this.dma_autoinit = false;
+});
+
+// DSP identification
+register_dsp_command([0xE0], 1, function()
+{
+    this.read_buffer.clear();
+    this.read_buffer.push(~this.write_buffer.shift());
+});
+
+// Get DSP version number.
+register_dsp_command([0xE1], 0, function()
+{
+    this.read_buffer.clear();
+    this.read_buffer.push(4);
+    this.read_buffer.push(5);
+});
+
+// DMA identification - based completely off dosbox
+/** @const */ var SB_E2_SIGNS = [1, -1, -1, 1];
+/** @const */ var SB_E2_EXTRA = [-106, 165, -151, 90];
+register_dsp_command([0xE2], 1, function()
+{
+    var input = this.write_buffer.shift();
+    dbg_log("e2 dma identification. (not understood) input: " + input, LOG_SB16);
+
+    this.e2_value += (input & 0x01) * SB_E2_SIGNS[(this.e2_count + 0) % 4];
+    this.e2_value += (input & 0x02) * SB_E2_SIGNS[(this.e2_count + 2) % 4];
+    this.e2_value += (input & 0x04) * SB_E2_SIGNS[(this.e2_count + 1) % 4];
+    this.e2_value += (input & 0x08) * SB_E2_SIGNS[(this.e2_count + 3) % 4];
+    this.e2_value += (input & 0x10) * SB_E2_SIGNS[(this.e2_count + 2) % 4];
+    this.e2_value += (input & 0x20) * SB_E2_SIGNS[(this.e2_count + 0) % 4];
+    this.e2_value += (input & 0x40) * SB_E2_SIGNS[(this.e2_count + 3) % 4];
+    this.e2_value += (input & 0x80) * SB_E2_SIGNS[(this.e2_count + 1) % 4];
+
+    this.e2_value += SB_E2_EXTRA[this.e2_count % 4];
+
+    this.e2_count++;
+    this.e2_value %= 256;
+
+    this.dma.on_unmask(this.dma_channel_8bit, () =>
```

Comment:
Functions cannot be stored in the state since they can't be serialised. Instead of storing `on_unmask` a function in `on_unmask`, can we store the variables that are captured in the closure of this inline function, and move it to the prototype of SB16?